### ISSUE TITLE
pos_weight should be a vector instead of a scalar

### ISSUE THREAD

**Issue by seermer** [ultralytics/yolov5#8749]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

In the current implementation, pos_weight is set to a scalar (default 1.0).
However, according to the [PyTorch official documentation](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html), quote:

> pos_weight ([Tensor](https://pytorch.org/docs/stable/tensors.html#torch.Tensor), optional) – a weight of positive examples. Must be a vector with a length equal to the number of classes.
also a given example:

> For example, if a dataset contains 100 positive and 300 negative examples of a single class, then pos_weight for the class should be equal to 300/100=3. The loss would act as if the dataset contains 3 x 100=300 positive examples.

therefore, pos_weight should be basically class weight for classification loss.

### Environment

- YOLOv5 6.1

### Minimal Reproducible Example

current implementation passes only a single element vector (a scalar) no matter how many classes it has;
`BCEcls = nn.BCEWithLogitsLoss(pos_weight=torch.tensor([h['cls_pw']], device=device))`

### Additional

related issue: https://github.com/ultralytics/yolov5/discussions/5604
in the above issue, pos_weight is misinterpreted as a scalar that is different from the class weight

possible fix:
1. change default hyp yaml to be a list instead of a scalar (I personally don't like this because if there are too many classes, this list will be super long)
2. change default cls loss pos_weight to torch.ones([nc]) * hyp['cls_pw']   (this sets all class to a same class weight, and hyp['cls_pw'] can stay as a scalar)
3. ~~change default cls loss pos_weight to model.class_weights * hyp['cls_pw']   (I personally prefer this since class_weights are computed already, we can use it easily, and the hyp['cls_pw'] will stay as a scalar that scales all the class weights )~~ **edit**: I actually just realized that the computed model.class_weights is a different thing from the PyTorch doc, so if we want to match what is suggested in PyTorch doc, we might need to compute a new class weight, for example, we can define it as **(total samples / nc) / samples per class * hyp['cls_pw']**. In this way, all classes will be trained as if there are **(total samples / nc)** samples and scaled by hyp['cls_pw'].

**edit**: please let me know if I misunderstood anything. Thanks.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@seermer thanks for the feedback! Yes #3 is the best solution I think, or perhaps #3 multiplied by the hyp pw scalar?

**Comment by seermer** [comment]
> @seermer thanks for the feedback! Yes #3 is the best solution I think, or perhaps #3 multiplied by the hyp pw scalar?

yes I think so, 3 probably is the best, but 3 probably doesn't work well with the --img-weights argument since it is changing the sampling frequency, so maybe we can do 2 when --img-weights is enabled, and use 3 otherwise

**Comment by glenn-jocher** [comment]
@seermer oh good point! Can you please submit a PR for this change? Would be interested to see what it does on COCO training.

**Comment by seermer** [comment]
I have just submitted a PR, its my first time doing a PR, hopefully, I'm doing things correctly

**Comment by glenn-jocher** [comment]
@seermer thank you! I will take a look when I have some time, extremely busy unfortunately.
